core/src/main/java/org/apache/iceberg/hadoop/HadoopTableOperations.java

```diff
@@ -368,58 +431,63 @@ private void renameToFinal(FileSystem fs, Path src, Path dst, int nextVersion) {
       if (fs.exists(dst)) {
         throw new CommitFailedException("Version %d already exists: %s", nextVersion, dst);
       }
-
-      if (!fs.rename(src, dst)) {
-        CommitFailedException cfe =
-            new CommitFailedException("Failed to commit changes using rename: %s", dst);
-        RuntimeException re = tryDelete(src);
-        if (re != null) {
-          cfe.addSuppressed(re);
-        }
-        throw cfe;
-      }
-    } catch (IOException e) {
-      CommitFailedException cfe =
-          new CommitFailedException(e, "Failed to commit changes using rename: %s", dst);
-      RuntimeException re = tryDelete(src);
-      if (re != null) {
-        cfe.addSuppressed(re);
+      if (!nextVersionIsLatest(nextVersion, fs)) {
```

Comment:
I don't think we need the extra check here. Won't we fail on the rename anyway?

Reply:
It is possible that rename will work. This is because the old version may have already been removed. So I introduced a check. This scenario occurs mostly in streaming write scenarios.

Reply:
If the locking is implemented correctly, then this should not happen if I follow the logical correctly. This operation will list the prefix, which can be costly:
- Doing additional operations to the filesystem/object-store
- Keeping the lock for a longer time than needed, slowing down the commit process

Reply:
@Fokko  For object stores, fs.rename may leave an incomplete dst file. This is because many object stores implement the rename operation as write(). As a result, all of the logic we have implemented in hadooptableOptions is not really applicable to object stores. 
For filesystems that support atomic operations, the list operation should take an acceptable amount of time if the number of metadata files is not very large.

Reply:
@Fokko 
Also, this check I admit does not solve the problem in all scenarios at the moment, a better approach would be to call fs.rename successfully and then check that it's commits are normal. This check just intended to fail the client as soon as possible. Because there are a lot of clients running flink/kafka-stream to write to flink at the same time they are running the compaction task. Implementing fast failures is more friendly to them.